### PR TITLE
FPS Widget

### PR DESCRIPTION
**Pull request by Mr-Anyone**
## High Level Overview

I added a fps widget. This doesn't seem to break anything from the look of things. I've attached an image of what it looks like. 

Screenshot from 2024-02-10 12-39-33

Screenshot from 2024-02-10 12-18-53

### Testing Done

Works on my end, and I think it work.  

### Resolved Issues

Added an FPS Widget.

### Review Checklist

- [x]  **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x]  **Remove all commented out code**
- [x]  **Remove extra print statements**: for example, those just used for testing
- [x]  **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/thunderscope/common/fps_widget.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    def refresh(self):
+        """
+        Updating the fps based on the frametime_counter
+        """
+        # updating every self.update_delta time, which is likely 0.5 seconds
+        if time.time() - self.last_update_time < self.update_delta:
+            return
+
+        # pull the data from frametime counter, the * 1000 is to display data ms instead of s
+        buffer_frametime = self.buffer_counter.get_last_frametime() * 1000
```
Comment (nimazareian) [inline src/software/thunderscope/common/fps_widget.py]: nit: We should be using these constants for converting between units

https://github.com/UBC-Thunderbots/Software/blob/5e7209758bcf150b55349e35d286c5e650d3fec9/src/shared/constants.h#L96-L99

File: src/software/thunderscope/gl/helpers/extended_gl_view_widget.py [hunk 2]
```diff
@@ -50,7 +51,7 @@ class ExtendedGLViewWidget(GLViewWidget):
     # (detect_mouse_movement_in_scene must be enabled for this signal to be emitted)
     mouse_in_scene_moved_signal = QtCore.pyqtSignal(MouseInSceneEvent)
 
-    def __init__(self) -> None:
+    def __init__(self, bufferswap_counter: FrameTimeCounter = None) -> None:
         """Initialize the ExtendedGLViewWidget"""
```
Comment (nimazareian) [inline src/software/thunderscope/gl/helpers/extended_gl_view_widget.py]: nit: the doc should be updated to include the new arg

File: src/software/thunderscope/common/frametime_counter.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import time
+from PyQt6.QtWidgets import *
+
+
+class FrameTimeCounter:
+    """
+    Frametimecounter is basically just a list that stores the time difference between each consecutaive function call.
```
Comment (itsarune) [inline src/software/thunderscope/common/frametime_counter.py]:
```suggestion
    FrameTimeCounter is basically just a list that stores the time difference between each consecutive function call.
```

File: src/software/thunderscope/thunderscope_config.py [hunk 4]
```diff
@@ -289,6 +313,13 @@ def configure_two_ai_gamecontroller_view(
     # Must be called before widgets are initialized below
     initialize_application()
 
+    # setup frametime counter
+    blue_refresh_func_frametime_counter = FrameTimeCounter()
+    blue_buffer_func_frametime_counter = FrameTimeCounter()
```
Comment (itsarune) [inline src/software/thunderscope/thunderscope_config.py:318]: what is the buffer_func_frametime_counter tracking?

File: src/software/thunderscope/thunderscope_types.py [hunk 5]
```diff
@@ -83,7 +84,11 @@ class TScopeQTTab(TScopeTab):
     ]  # Mapping of widget names to refresh functions
 
     def __init__(
-        self, name: str, key: TabNames, widgets: Sequence[TScopeWidget]
+        self,
+        name: str,
+        key: TabNames,
+        widgets: Sequence[TScopeWidget],
+        refresh_func_counter: FrameTimeCounter = None,
```
Comment (itsarune) [inline src/software/thunderscope/thunderscope_types.py:91]: add docs for the new arg that you added

File: src/software/thunderscope/thunderscope_config.py (hunk 4 above)
Comment (itsarune) [reply src/software/thunderscope/thunderscope_config.py:318]: Or I guess I'm confused why we have two counters for the refresh and buffer swap. what does one metric give us that another one doesnt?

**Review by itsarune — COMMENTED** [review]
nice work!

I find it a bit confusing trying to interpret the FPS widget with the new data that you're showing. Maybe you should consider simplifying the data output and just show what's the most important thing we need to understand thunderscope performance

**Comment by Mr-Anyone** [comment]
Screenshot from 2024-02-21 18-01-27
This is the new look.

File: src/software/thunderscope/common/frametime_counter.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        self.datapoints = []  # stores the timeframe of every data cycle
+        self.previous_timestamp = time.time()
+
+    def add_one_datapoint(self):
+        """
+        Save the time difference between each consecutative function call.
+        """
+        current_time = time.time()
+        time_difference = current_time - self.previous_timestamp
+        self.datapoints.append(time_difference)
```
Comment (nimazareian) [inline src/software/thunderscope/common/frametime_counter.py]: `datapoints` is an endlessly growing array. You should consider using a [`collections.deque`](http://docs.python.org/library/collections.html#collections.deque) instead with a limited size

File: src/software/thunderscope/common/frametime_counter.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import time
+from PyQt6.QtWidgets import *
+
+
+class FrameTimeCounter:
+    """
+    FrameTimeCounver is basically just a list that stores the time difference between each consecutive function call.
+    From that, it calcualtes the frametime of each function call.
```
Comment (nimazareian) [inline src/software/thunderscope/common/frametime_counter.py]: nit:
```suggestion
    From that, it calculates the frametime of each function call.
```

File: src/software/thunderscope/common/frametime_counter.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self) -> None:
+        """
+        Initialized framtime counter
+        """
+        self.datapoints = []  # stores the timeframe of every data cycle
+        self.previous_timestamp = time.time()
+
+    def add_one_datapoint(self):
+        """
+        Save the time difference between each consecutative function call.
```
Comment (nimazareian) [inline src/software/thunderscope/common/frametime_counter.py]: nit:
```suggestion
        Save the time difference between each consecutive function call.
```

File: src/software/thunderscope/common/frametime_counter.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        :return: the average of the entire list, and -1 if the list is empty
+        """
+        if len(self.datapoints) == 0:
+            return -1
+
+        return sum(self.datapoints) / len(self.datapoints)
+
+    def get_average_last_30(self):
+        """
+        Averaget the last 30 frametime
```
Comment (nimazareian) [inline src/software/thunderscope/common/frametime_counter.py]: nit:
```suggestion
        Average the last 30 frametime
```

File: src/software/thunderscope/common/fps_widget.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    the refresh function in tab.
+
+    Also, this widget update every 0.5 seconds
+    """
+
+    def __init__(
+        self,
+        buffer_counter: FrameTimeCounter,
+        refresh_counter: FrameTimeCounter,
+        update_delta=0.5,
```
Comment (nimazareian) [inline src/software/thunderscope/common/fps_widget.py]: nit:
```suggestion
        update_delta_sec=0.5,
```

File: src/software/thunderscope/common/fps_widget.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+        :buffer_counter: a frametime counter at the GLWidget Widget
+        :refresh_counter: a frametime counter at the refresh function
+        """
+        super().__init__()
+        self.buffer_counter = buffer_counter
+        self.refresh_counter = refresh_counter
+
+        self.fps_label = QLabel("some string to be show")
+        self.fps_label.setText("some fps: ")
```
Comment (nimazareian) [inline src/software/thunderscope/common/fps_widget.py]: it doesn't seem like this variable is being used?

File: src/software/thunderscope/common/fps_widget.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        self.vertical_layout = QVBoxLayout()
+
+        self.buffertime_table = QTableWidget(3, 2)
+        self.buffertime_table.setHorizontalHeaderLabels(["Frametime (ms)", "FPS"])
+        self.buffertime_table.setVerticalHeaderLabels(["Recent", "Last 30", "All"])
+        self.refresh_function_table = QTableWidget(3, 2)
+        self.refresh_function_table.setHorizontalHeaderLabels(["Frametime (ms)", "FPS"])
+        self.refresh_function_table.setVerticalHeaderLabels(
+            ["Recent", "Last 30", "All"]
+        )
```
Comment (nimazareian) [inline src/software/thunderscope/common/fps_widget.py:42]: You should consider to resize the columns so the titles show up better
```py
        self.buffertime_table.resizeColumnsToContents()
        self.refresh_function_table.resizeColumnsToContents()
```

File: src/software/thunderscope/gl/helpers/extended_gl_view_widget.py [hunk 13]
```diff
@@ -66,6 +70,19 @@ def __init__(self) -> None:
...
+        if self.bufferswap_counter == None:
+            self.bufferswap_counter = FrameTimeCounter()
+        self.frameSwapped.connect(self.frameswap_callback)
+
+    def frameswap_callback(self):
+        """
+        addding a frameswap callback
+        """
+        # adding a a
+        self.bufferswap_counter.add_one_datapoint()
```
Comment (nimazareian) [inline src/software/thunderscope/gl/helpers/extended_gl_view_widget.py]: nit: This comment seems to be incomplete

File: src/software/thunderscope/gl/helpers/extended_gl_view_widget.py [hunk 14]
```diff
@@ -50,8 +51,11 @@ class ExtendedGLViewWidget(GLViewWidget):
     # (detect_mouse_movement_in_scene must be enabled for this signal to be emitted)
     mouse_in_scene_moved_signal = QtCore.pyqtSignal(MouseInSceneEvent)
 
-    def __init__(self) -> None:
-        """Initialize the ExtendedGLViewWidget"""
+    def __init__(self, bufferswap_counter: FrameTimeCounter = None) -> None:
+        """
+        Initialize the ExtendedGLViewWidget
+
+        """
```
Comment (nimazareian) [inline src/software/thunderscope/gl/helpers/extended_gl_view_widget.py]: nit: `bufferswap_counter` is missing from the function docs

File: src/software/thunderscope/thunderscope_types.py [hunk 15]
```diff
@@ -136,9 +153,13 @@ def refresh(self) -> None:
...
-        # only refresh dock that are visible
-        if not self.dock_area.isVisible():
-            return
+
+        self.refresh_func_counter.add_one_datapoint()
+
+        for refresh_func in self.refresh_functions.values():
+            # only refresh dock that are visible
+            if not self.dock_area.isVisible():
+                return
```
Comment (nimazareian) [inline src/software/thunderscope/thunderscope_types.py]: is this necessary? Is the next for loop not handling this?

**Review by nimazareian — COMMENTED** [review]
The new layout looks much better! Good work! Just had a couple of nits

File: src/software/thunderscope/common/frametime_counter.py (hunk 9 above)
Comment (Mr-Anyone) [reply src/software/thunderscope/common/frametime_counter.py]: Done

**Comment by nimazareian** [comment]
You should pull from upstream master to fix the broken CI

**Review by nimazareian — APPROVED** [review]
:100:

**Review by williamckha — APPROVED** [review]

**Pull request merged**